api/tests/test_bot_server.py

```diff
@@ -0,0 +1,63 @@
+import zulip.bot_server
+import unittest
+import mock
+import json
+from typing import Any
+
+class BotServerTests(unittest.TestCase):
+    class MockMessageHandler(object):
+        def handle_message(self, message, bot_handler, state_handler):
+            # type: (Any, Any, Any) -> None
+            assert message == {'key': "test message"}
+
+    class MockLibModule(object):
+        def handler_class(self):
+            # type: () -> Any
+            return BotServerTests.MockMessageHandler()
+
+    def setUp(self):
+        # type: () -> None
+        zulip.bot_server.app.testing = True
+        self.app = zulip.bot_server.app.test_client()
+
+    @mock.patch('zulip.bot_server.BotHandlerApi')
+    def test_successful_request(self, mock_BotHandlerApi):
+        # type: (mock.Mock) -> None
+        zulip.bot_server.available_bots = ['testbot']
+        zulip.bot_server.bots_config = {
+            'testbot': {
+                'email': 'testbot-bot@zulip.com',
+                'key': '123456789qwertyuiop',
+                'site': 'http://localhost',
+            }
+        }
+        zulip.bot_server.bots_lib_module['testbot'] = BotServerTests.MockLibModule()
+        response = self.app.post('/bots/testbot', data=json.dumps(dict(message={'key': "test message"})))
+
+        assert response.status_code >= 200 and response.status_code < 300
+        assert mock_BotHandlerApi.called
+
+    def test_bot_not_supported(self):
+        # type: () -> None
+        if 'not_supported_bot' in zulip.bot_server.available_bots:
+            zulip.bot_server.available_bots.remove('not_supported_bot')
+        response = self.app.post('/bots/not_supported_bot', data=json.dumps(dict(message={'key': "test message"})))
+
+        assert response.status_code >= 400 and response.status_code < 500
+
+    def test_wrong_bot_credentials(self):
+        # type: () -> None
+        zulip.bot_server.available_bots = ['testbot']
+        zulip.bot_server.bots_config = {
+            'testbot': {
+                'email': 'testbot-bot@zulip.com',
+                'key': '123456789qwertyuiop',
+                'site': 'http://localhost',
+            }
+        }
+        response = self.app.post('/bots/testbot', data=json.dumps(dict(message={'key': "test message"})))
+
+        assert response.status_code >= 400 and response.status_code < 500
+
+if __name__ == '__main__':
+    unittest.main()
```

Comment:
Looks great, but I was wondering, is it worth it to create a test suite class for this, so the logic for setting `bots_config`, `response = ...` and asserting `status_code` can be wrapped by an assertion helper called something like `assertBotServerMessage('testbot', config, message)`. I think it would be a great addition, especially if we're going to add new features to this server. What do you think?

Although I'm not entirely sure what Tim meant by making a "test suite" for this. :)